File: arrays/shortest-distance-to-a-character/shortest-distance-to-a-character.ts
function shortestToChar(s: string, c: string): number[] {
  let n: number = s.length;
  const res: number[] = [];
  let prev: number = -n;

  for (let i = 0; i < n; i++) {
    if (s[i] === c) {
      prev = i;
    }

    res[i] = i - prev;
  }

  for (let i = prev - 1; i >= 0; i--) {
    if (s[i] === c) {
      prev = i;
    }

    res[i] = Math.min(res[i], prev - i);
  }

  return res;
}

File: arrays/shortest-distance-to-a-character/shortest-distance-to-a-character.java
class Solution {
  public int[] shortestToChar(String s, char c) {
    int n = s.length();
    int prev = -n;
    int[] res = new int[n];

    for (int i = 0; i < n; i++) {
      if (s.charAt(i) == c) {
        prev = i;
      }
      res[i] = i - prev;
    }

    for (int i = prev - 1; i >= 0; i--) {
      if (s.charAt(i) == c) {
        prev = i;
      }
      res[i] = Math.min(res[i], prev - i);
    }

    return res;
  }
}
